src/components/App/AppLoader.js

```diff
@@ -0,0 +1,128 @@
+import React, { useState, useEffect } from 'react'
+import PropTypes from 'prop-types'
+import LoadingScreen from '../LoadingScreen/LoadingScreen'
+
+const INTERNAL_APPS = ['home', 'permissions', 'apps', 'settings']
+
+const LOADING_ORG = 'Loading organization…'
+const LOADING_APPS = 'Loading apps…'
+const LOADING_APP = 'Loading {APP}…'
+const LOADING_READY = 'Ready.'
+
+// Pick the loading steps depending on what
+// is being loaded (internal vs.  external app)
+const LOADING_STEPS_INTERNAL = {
+  steps: [LOADING_ORG, LOADING_APPS, LOADING_READY],
+  stepIndex: ({ daoLoading, appsLoading }) =>
+    // pick the first truthy item as the current step
+    [daoLoading, appsLoading, true].findIndex(v => v),
```

Comment:
Wondering if it makes sense to block the user from the internal apps until the external apps are loaded? I think having access to the internal apps makes sense even if it's information is incomplete and we're still waiting to load the apps.

It would also make the app feel like it loads a bit faster, since the apps can take some time to load initially.

Reply:
Yes! That’s the reason why I added these dynamic steps, even if there are really only two modes for now. The idea was to reset the inconsistent loading states that were a added at the last minute (home / apps / settings / permissions), then progressively replace them by consistent / redesigned ones.

But I realize now that settings might be problematic, as the network part haven’t been migrated to the “Preferences” screen, and we might want to access them if apps are loading too slowly.

So I think we could load these units for now:

- Home: [dao, apps]
- App: [dao, apps, app]
- Permissions: [dao, apps]
- App Center: [dao, apps]
- Settings: [dao] (until the network settings get moved to “Preferences”).

Then we could start to display some apps earlier where it makes sense, as in-app loading states are getting redesigned (e.g. for the permissions refresh).

What do you think?

Reply:
Makes sense :).